src/org/mozilla/javascript/NativeConsole.java

```diff
@@ -0,0 +1,437 @@
+        NativeConsole obj = new NativeConsole(printer);
+        obj.activatePrototypeMap(MAX_ID);
+        obj.setPrototype(getObjectPrototype(scope));
+        obj.setParentScope(scope);
+        if (sealed) {
+            obj.sealObject();
+        }
+        ScriptableObject.defineProperty(scope, "console", obj, ScriptableObject.DONTENUM);
+    }
+
+    private NativeConsole(ConsolePrinter printer) {
+        this.printer = printer;
+    }
+
+    @Override
+    public String getClassName() {
+        return "Console";
+    }
+
+    @Override
+    protected void initPrototypeId(int id) {
+        if (id > LAST_METHOD_ID) {
+            throw new IllegalStateException(String.valueOf(id));
+        }
+
+        String name;
+        int arity;
+        switch (id) {
+            case Id_toSource:
+                arity = 0;
+                name = "toSource";
+                break;
+            case Id_trace:
+                arity = 1;
+                name = "trace";
+                break;
+            case Id_debug:
+                arity = 1;
+                name = "debug";
+                break;
+            case Id_log:
+                arity = 1;
+                name = "log";
+                break;
+            case Id_info:
+                arity = 1;
+                name = "info";
+                break;
+            case Id_warn:
+                arity = 1;
+                name = "warn";
+                break;
+            case Id_error:
+                arity = 1;
+                name = "error";
+                break;
+            case Id_assert:
+                arity = 2;
+                name = "assert";
+                break;
+            case Id_count:
+                arity = 1;
+                name = "count";
+                break;
+            case Id_countReset:
+                arity = 1;
+                name = "countReset";
+                break;
+            case Id_time:
+                arity = 1;
+                name = "time";
+                break;
+            case Id_timeEnd:
+                arity = 1;
+                name = "timeEnd";
+                break;
+            case Id_timeLog:
+                arity = 2;
+                name = "timeLog";
+                break;
+            default:
+                throw new IllegalStateException(String.valueOf(id));
+        }
+        initPrototypeMethod(CONSOLE_TAG, id, name, arity);
+    }
+
+    @Override
+    public Object execIdCall(
+            IdFunctionObject f, Context cx, Scriptable scope, Scriptable thisObj, Object[] args) {
+        if (!f.hasTag(CONSOLE_TAG)) {
+            return super.execIdCall(f, cx, scope, thisObj, args);
+        }
+
+        int methodId = f.methodId();
+        switch (methodId) {
+            case Id_toSource:
+                return "Console";
+
+            case Id_trace:
+                printer.print(cx, scope, Level.TRACE, args);
```

Comment:
Shouldn't console.trace() automatically send the current stacktrace as the first entry in the args object[] to the printer?

Reply:
How about the latest code which passes `ScriptStackElement[] stack` before `Object[] args`.